mapreduce/src/main/java/uk/gov/gchq/palisade/PalisadeRecordReader.java

```diff
@@ -0,0 +1,229 @@
+import uk.gov.gchq.palisade.service.request.DataRequestResponse;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The main {@link RecordReader} class for Palisade MapReduce clients. This class implements the logic for connecting to
+ * a data service and retrieving the results from the returned stream. It will contact the necessary data service for
+ * each resource in turn from the {@link PalisadeInputSplit} provided to it. Clients can retrieve the current resource
+ * being processed by calling {@link PalisadeRecordReader#getCurrentKey()}. Therefore, in the client MapReduce code, the
+ * key will become the current resource being processed and the value will become the current item from that resource.
+ *
+ * @param <V> the value type which will be de-serialised from the resources this input split is processing
+ */
+public class PalisadeRecordReader<V> extends RecordReader<Resource, V> {
+    /*
+     * Internal class description:
+     * This class currently requests each Resource from its data service sequentially. We avoid launching all the
+     * requests to the data service(s) in parallel because Hadoop's processing of tasks in an individual map task is necessarily
+     * serial. If we launch multiple requests for data in parallel, but Hadoop/the user's MapReduce job spends a long
+     * time processing the first Resource(s), then the data services waiting to send the ones later in the list may timeout.
+     * Thus, we only make the request to the data service responsible for an individual resource when we need it. This
+     * change in the future and SHOULD NOT be relied upon in any implementation decisions.
+     *
+     * In order to do this, we create a DataRequestResponse for each Resource and send it to the data service created by
+     * the corresponding ConnectionDetail object.
+     */
+
+    /**
+     * The request that is being processed in this task.
+     */
+    private DataRequestResponse resourceDetails;
+
+    /**
+     * Iterates through the resources to be processed.
+     */
+    private Iterator<Map.Entry<Resource, ConnectionDetail>> resIt;
+
+    /**
+     * Value supplier.
+     */
+    private Iterator<V> itemIt;
+
+    /**
+     * The current Palisade resource being processed.
+     */
+    private Resource currentKey;
+
+    /**
+     * The current value in a resource.
+     */
+    private V currentValue;
+
+    /**
+     * The Palisade serialiser that will be used to decode each item from the resources being processed.
+     */
+    private Serialiser<Object, V> serialiser;
+
+    /**
+     * Count of number of resources already processed, used for Haodop progress monitoring.
+     */
+    private int processed;
+
+    /**
+     * No-arg constructor.
+     */
+    public PalisadeRecordReader() {
+    }
+
+    //TODO: error handling
```

Comment:
Please could you create a GitHub issue in to do this error handling so it doesn't get forgotten about.

Reply:
Done